pytorch_lightning/plugins/training_type/deepspeed.py

```diff
@@ -0,0 +1,321 @@
+                This is a speed optimization when training across multiple GPUs/machines. (default: True)
+
+            allgather_partitions: All gather updated parameters at the end of training step,
+                instead of using a series of broadcast collectives (default: True)
+
+            reduce_scatter: Use reduce/scatter instead of allreduce to average gradients (default:True)
+
+            allgather_bucket_size: Number of elements to allgather at once.
+                Used to limit the memory required for larger model sizes, with a tradeoff with speed. (default: 2e8)
+
+            reduce_bucket_size: Number of elements to reduce at once.
+                Used to limit the memory required for larger model sizes, with a tradeoff with speed (default: 2e8)
+
+            zero_allow_untested_optimizer: Allow untested optimizers to be used with ZeRO. Currently only Adam is a
+                DeepSpeed supported optimizer when using ZeRO (default: True)
+
+            config: Pass in a deepspeed formatted config dict,
+                or path to a deepspeed config: https://www.deepspeed.ai/docs/config-json.
+                All defaults will be ignored if a config is passed in. (Default: ``None``)
+
+            logging_level: Set logging level for deepspeed. (Default: ``logging.WARN``)
+
+        """
+        super().__init__(
+            parallel_devices=parallel_devices, num_nodes=num_nodes, cluster_environment=cluster_environment
+        )
+        self.config = self._load_config(config)
+        if self.config is None:
+            # User has not overridden config, set defaults
+            self.config = self._create_default_config(
+                zero_optimization,
+                zero_allow_untested_optimizer,
+                stage=stage,
+                cpu_offload=cpu_offload,
+                contiguous_gradients=contiguous_gradients,
+                overlap_comm=overlap_comm,
+                allgather_partitions=allgather_partitions,
+                reduce_scatter=reduce_scatter,
+                allgather_bucket_size=allgather_bucket_size,
+                reduce_bucket_size=reduce_bucket_size
+            )
+        self._config_initialized = False
+        deepspeed.utils.logging.logger.setLevel(logging_level)
+
+    def _load_config(self, config):
+        if config is None and self.DEEPSPEED_ENV_VAR in os.environ:
+            rank_zero_info(f"Loading DeepSpeed config from set {self.DEEPSPEED_ENV_VAR} environment variable")
+            config = os.environ[self.DEEPSPEED_ENV_VAR]
+        if isinstance(config, str) or isinstance(config, Path):
+            if os.path.exists(config):
+                with open(config) as f:
+                    config = json.load(f)
+            else:
+                raise MisconfigurationException(
+                    f"You passed in a path to a DeepSpeed config but the path does not exist: {config}"
+                )
+        return config
+
+    def pre_training(self):
```

Comment:
This is also called during `trainer.test` right?

Reply:
Yes, name has changed to `pre_dispatch` which is clearer